File: python/v1.py
from utils import measurements_file, round_half_up

from typing import TypedDict


class CityReading(TypedDict):
  minimum: int
  maximum: int
  total: int
  count: int


if __name__ == "__main__":
  results: dict[str, CityReading] = {}

  with open(measurements_file) as measurements_file:
    for line in measurements_file:
      city, reading_str = line.strip().split(";")
      reading = float(reading_str)

      current = results.get(city, None)

      if current is None:
        results[city] = CityReading(
          minimum=reading,
          maximum=reading,
          total=reading,
          count=1,
        )
      else:
        current["minimum"] = min(current["minimum"], reading)
        current["maximum"] = max(current["maximum"], reading)
        current["total"] += reading
        current["count"] += 1
        results[city] = current

  cities = []
  for city in sorted(results.keys()):
    values = results[city]
    ans_min = round_half_up(values["minimum"])
    ans_avg = round_half_up(values["total"] / values["count"])
    ans_max = round_half_up(values["maximum"])
    cities.append(
      "%s=%.1f/%.1f/%.1f"
      % (
        city,
        ans_min,
        ans_avg,
        ans_max,
      )
    )

  print("{" + ", ".join(cities) + "}")

File: node/v1.ts
import { createInterface } from "readline";
import { createReadStream, read } from "fs";
import { filePath, roundHalfUp } from "./utils";

const fileStream = createReadStream(filePath);

const input = createInterface({
  input: fileStream,
});

type CityReading = { min: number; max: number; total: number; count: number };

const readings = new Map<string, CityReading>();

const parseLine = (line: string) => {
  const parts = line.split(";");
  const city = parts[0];
  const reading = Number(parts[1]);

  const current = readings.get(city);

  if (!current) {
    readings.set(city, {
      min: reading,
      max: reading,
      total: reading,
      count: 1,
    });
  } else {
    readings.set(city, {
      min: Math.min(current.min, reading),
      max: Math.max(current.max, reading),
      total: current.count + reading,
      count: current.count + 1,
    });
  }
};

const processResults = () => {
  const results = [...readings.keys()].sort().map((city) => {
    const values = readings.get(city) as CityReading;

    const min = roundHalfUp(values.min);
    const max = roundHalfUp(values.max);
    const avg = roundHalfUp(values.total / values.count);

    return `${city}=${min.toFixed(1)}/${avg.toFixed(1)}/${max.toFixed(1)}`;
  });

  console.log(`{${results.join(", ")}}`);
};

input.on("line", parseLine);
input.on("close", processResults);
